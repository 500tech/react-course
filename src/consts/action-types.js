export const ADD_RECIPE = 'ADD_RECIPE';
export const TOGGLE_RECIPE = 'TOGGLE_RECIPE';
export const SET_USER = 'SET_USER';