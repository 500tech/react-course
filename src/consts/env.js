export const BASE_URL = 'https://s3.amazonaws.com/500tech-shared/';
