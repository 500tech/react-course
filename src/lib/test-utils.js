export const mockResponse = (status, statusText, response) => {
  return new window.Response(response, {
    status: status,
    statusText: statusText,
    headers: {
      'Content-type': 'application/json'
    }
  });
};

const handleResponse = (mockedUrl, response) =>
  window.fetch = jest.fn().mockImplementation(url => {
    if (url === mockedUrl) {
      return response;
    }
    throw('Unknown URL: ' + url);
  });

export const mockFetch = (mockedUrl, status, data) =>
  handleResponse(
    mockedUrl,
    Promise.resolve(mockResponse(status, null, data)));

export const mockFetchError = (mockedUrl, state, error) =>
  handleResponse(
    mockedUrl,
    Promise.reject(mockResponse(state, error, '{}')));
