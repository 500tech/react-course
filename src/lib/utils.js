let id = 0;
export const getID = () => id += 1;
