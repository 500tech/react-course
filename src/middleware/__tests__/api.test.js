import apiMiddleware from '../api';
import { mockFetch, mockFetchError } from '../../lib/test-utils';
import { API } from '../../consts/action-types';
import { BASE_URL } from '../../consts/env';

const data = { title: 'hello' };

const apiAction = () => ({
  type: API,
  payload: {
    success: 'SET_DATA',
    url: 'fake.json'
  }
});

describe("api middleware", () => {
  let next, dispatch, middleware, dispatchCalls, nextCalls, getState;

  const getStateMock = () => getState ? getState() : {};

  beforeEach(() => {
    getState      = null;
    next          = jest.fn();
    dispatch      = jest.fn();
    dispatchCalls = dispatch.mock.calls;
    nextCalls     = next.mock.calls;
    middleware    = apiMiddleware({ dispatch, getState: getStateMock })(next);
  });

  it('should ignore non API actions', () => {
    const sampleAction = { type: 'SAMPLE_ACTION' };

    middleware(sampleAction);

    expect(dispatchCalls.length).toBe(0);
    expect(nextCalls).toEqual([[sampleAction]]);
  });

  describe('success', () => {
    beforeEach(() => mockFetch(BASE_URL + 'fake.json', 200, JSON.stringify(data)));

    it('should dispatch SET_DATA', () => middleware(apiAction()).then(() =>
        expect(dispatchCalls[0][0]).toEqual({ type: 'SET_DATA', payload: data })));
  });

  describe('error', () => {
    beforeEach(() => mockFetchError(BASE_URL + 'fake.json', 404, 'Not found'));

    it('should dispatch error', () => middleware(apiAction()).then(() =>
      expect(dispatchCalls[0]).toMatchSnapshot()));
  });
});
