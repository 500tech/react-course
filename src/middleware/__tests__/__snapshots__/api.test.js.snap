// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api middleware error should dispatch error 1`] = `
Array [
  Object {
    "error": Response {
      "_bodyInit": "{}",
      "_bodyText": "{}",
      "headers": Headers {
        "map": Object {
          "content-type": "application/json",
        },
      },
      "ok": false,
      "status": 404,
      "statusText": "Not found",
      "type": "default",
      "url": "",
    },
    "type": "API_ERROR",
  },
]
`;
